content/blog/using-the-gdb4hpc-debugger-at-scale-with-a-cuda-mpi-hpc-application.md

```diff
@@ -0,0 +1,545 @@
+nvcc to use the `cc` compiler to do the actual compilation of the generated
+C/C++ code. When the Cray CCE module is loaded, cc is the Cray C/C++ compiler.
+
+`-gencode` tells nvcc which GPU architecture to target. The actual value for
+this argument will vary based on which GPU you are targeting. Refer to the
+[NVIDIA compute capabilities table](https://developer.nvidia.com/cuda-gpus) to
+find the compute capability of your GPU. For example, the A100 GPU that I am
+targeting while writing this tutorial has a compute capability of 8.0. Take the compute
+capability of your GPU and pass it to nvcc via the `-gencode` flag by
+removing the decimal point and embedding it in an `arch=compute_XX,code=sm_XX`
+pair. For my A100 example, the complete option is `-gencode compute_80,code=sm_80`.
+
+`simpleMI.cu` is the input file and `-o simpleMPI_gpu.o` tells `nvcc` to call
+the output file `simpleMPI_gpu.o`.
+
+My complete nvcc invocation for an A100 GPU looks like this:
+
+```shell
+$ nvcc -g -G -c -ccbin=cc -gencode arch=compute_80,code=sm_80 simpleMPI.cu -o simpleMPI_gpu.o
+```
+
+Refer to the [`nvcc` user manual](https://docs.nvidia.com/cuda/cuda-compiler-driver-nvcc/index.html)
+and the [compilation section of the cuda-gdb documentation](https://docs.nvidia.com/cuda/cuda-gdb/index.html#compiling-the-application)
+for more information.
+
+#### Compile the .cpp File
+
+The simpleMPI.cpp file contains code that will be run on the CPU. It is the
+entry point to the program and contains the MPI operations. Use CC to compile
+it.
+
+```shell
+$ CC -g -O0 -c simpleMPI.cpp -o simpleMPI_cpu.o
+```
+
+The flags have the same meanings as in the nvcc compilation. The Cray CC
+compiler will automatically include the flags needed for building with MPI.
+
+#### Link the Application and Produce an Executable
+
+Finally, use CC to link the two object files into the final executable.
+
+```shell
+$ CC simpleMPI_gpu.o simpleMPI_cpu.o -o simpleMPI
+```
+
+#### Test The Application
+
+Test that the application was compiled correctly by running it on multiple
+nodes. For example, on a Slurm system:
+
+```shell
+$ srun -n8 -N2 --ntasks-per-node 4 --exclusive -p griz256 ./simpleMPI
+Running on 8 nodes
+Average of square roots is: 0.667337
+PASSED
+```
+
+Your srun command's flags may vary depending on your system. You should launch the
+job on nodes with your targeted GPU with one rank per GPU. See
+[the `srun` documentation](https://slurm.schedmd.com/srun.html)
+for how to achieve a correct job launch on your system.
+
+In my specific example:
+
+`-n8 -N2 --ntasks-per-node 4` tells Slurm to run 8 ranks across 2 nodes,
+splitting the job evenly by running 4 ranks on each node. In this example we
+are using nodes with 4 A100s each.
+
+`--exclusive` tells Slurm to run on nodes that aren't being used by anyone
+else, and to not let anyone else use them while running the job. Only one
+application can use each GPU at a time, so we need to make sure that we aren't
```

Comment:
I recommend to our writers to avoid the "Royal we". Stick with the pronouns you've used earlier. You've been giving instructions to "you", and that is perfect. In this case, you would just say "...so you need to make sure that you aren't given codes that have GPUs that are already in use."

Reply:
Thanks, this tutorial was actually converted from a previously written tutorial that used the royal we, and I thought I caught all the instances, hence the correctness up to this point. 😅